### PR TITLE
PT-1198: Implement Generic CRUD service

### PR DESCRIPTION
**Pull request by akak1977**
Generic CRUD/search services impementation.

https://virtocommerce.atlassian.net/browse/PT-1196

### CODE REVIEW

File: src/VirtoCommerce.Platform.Data/GenericCrud/CrudService.cs [hunk 1]
```diff
@@ -0,0 +1,193 @@
...
+
+        protected Task<TEntity[]> LoadEntities(IRepository repository, string[] ids)
+        {
+            return LoadEntities(repository, ids, "Full");
+        }
+
+        protected abstract void ValidateOnSave(TModel[] entitites);
+
+
+        protected abstract Task SoftDelete(IRepository repository, string[] ids);
```
Comment (eMazeika) [inline src/VirtoCommerce.Platform.Data/GenericCrud/CrudService.cs:95]: can change abstract to virtual with empty impl.?

File: src/VirtoCommerce.Platform.Data/GenericCrud/CrudService.cs [hunk 2]
```diff
@@ -0,0 +1,193 @@
...
+
+        protected abstract TEntity FromModel(TEntity entity, TModel model, PrimaryKeyResolvingMap pkMap);
+
+        protected abstract TModel ToModel(TEntity entity, TModel model);
+
+        protected abstract TModel Clone(TModel model);
+
+        protected abstract void Patch(TEntity sourceEntity, TEntity targetEntity);
+
+        protected abstract TModel PopulateModel(string responseGroup, TEntity entity, TModel model);
```
Comment (eMazeika) [inline src/VirtoCommerce.Platform.Data/GenericCrud/CrudService.cs:82]: can change abstract to virtual with simple impl.?

File: src/VirtoCommerce.Platform.Data/GenericCrud/CrudService.cs [hunk 3]
```diff
@@ -0,0 +1,193 @@
...
+
+
+        protected abstract Task<TEntity[]> LoadEntities(IRepository repository, string[] ids, string responseGroup);
+
+        protected Task<TEntity[]> LoadEntities(IRepository repository, string[] ids)
+        {
+            return LoadEntities(repository, ids, "Full");
+        }
+
+        protected abstract void ValidateOnSave(TModel[] entitites);
```
Comment (eMazeika) [inline src/VirtoCommerce.Platform.Data/GenericCrud/CrudService.cs:92]: can change abstract to virtual with empty impl.?

File: src/VirtoCommerce.Platform.Caching/GenericCrud/GenericCrudCachingRegion.cs [hunk 4]
```diff
@@ -0,0 +1,36 @@
...
+                throw new ArgumentNullException(nameof(entityIds));
+            }
+
+            var changeTokens = new List<IChangeToken> { CreateChangeToken() };
+            foreach (var entityId in entityIds)
+            {
+                changeTokens.Add(CreateChangeTokenForKey(entityId));
+            }
+            return new CompositeChangeToken(changeTokens);
+        }
```
Comment (Vectorfield4) [inline src/VirtoCommerce.Platform.Caching/GenericCrud/GenericCrudCachingRegion.cs:34]: No reason for exception.
```suggestion
        public static IChangeToken CreateChangeToken(IEnumerable<string> entityIds)
        {
            var changeTokens = new List<IChangeToken> { CreateChangeToken() };
            
            if (entityIds != null)
            {
                changeTokens.AddRange(entityIds.Select(entityId => CreateChangeTokenForKey(entityId)));
            }
            
            return new CompositeChangeToken(changeTokens);
        }
```

**Review by Vectorfield4 — CHANGES_REQUESTED** [review]

File: src/VirtoCommerce.Platform.Data/GenericCrud/SearchService.cs [hunk 5]
```diff
@@ -0,0 +1,91 @@
...
+                        }
+                        retVal.Results = (await _crudService.GetByIdsAsync(ids, criteria.ResponseGroup)).OrderBy(x => Array.IndexOf(ids, x.Id)).ToList();
+                    }
+
+                    if (needExecuteCount)
+                    {
+                        retVal.TotalCount = await query.CountAsync();
+                    }
+
+                    return retVal;
```
Comment (Vectorfield4) [inline src/VirtoCommerce.Platform.Data/GenericCrud/SearchService.cs:76]: Rename to results.

File: src/VirtoCommerce.Platform.Data/GenericCrud/SearchService.cs [hunk 6]
```diff
@@ -0,0 +1,91 @@
...
+            _platformMemoryCache = platformMemoryCache;
+            _repositoryFactory = repositoryFactory;
+            _crudService = crudService;
+        }
+
+
+        public virtual async Task<TResult> SearchAsync(TCriteria criteria)
+        {
+            var cacheKey = CacheKey.With(GetType(), nameof(SearchAsync), criteria.GetCacheKey());
+            var result = await _platformMemoryCache.GetOrCreateExclusiveAsync(cacheKey, async cacheEntry =>
```
Comment (Vectorfield4) [inline src/VirtoCommerce.Platform.Data/GenericCrud/SearchService.cs:38]: return

**Review by Vectorfield4 — CHANGES_REQUESTED** [review]

File: src/VirtoCommerce.Platform.Data/GenericCrud/CrudService.cs [hunk 7]
```diff
@@ -0,0 +1,194 @@
...
+            }
+        }
+
+        public virtual async Task DeleteAsync(string[] ids, bool softDelete = false)
+        {
+            var models = (await GetByIdsAsync(ids)).ToArray();
+
+            using (var repository = _repositoryFactory())
+            {
+                //Raise domain events before deletion
```
[the remark this answers is not in the document]
Comment (akak1977) [reply src/VirtoCommerce.Platform.Data/GenericCrud/CrudService.cs:160]: I don't know. It was there before.

File: src/VirtoCommerce.Platform.Caching/GenericCrud/GenericCrudCachingRegion.cs (hunk 4 above)
Comment (akak1977) [reply src/VirtoCommerce.Platform.Caching/GenericCrud/GenericCrudCachingRegion.cs]: Why not?
It was before.

File: src/VirtoCommerce.Platform.Data/GenericCrud/CrudService.cs (hunk 7 above)
[the remark this answers is not in the document]
Comment (Vectorfield4) [reply src/VirtoCommerce.Platform.Data/GenericCrud/CrudService.cs:160]: @tatarincev why should we send domain events of deletion before we actually commit it? Or maybe you can explain the logic of this code.

**Review by Vectorfield4 — APPROVED** [review]

File: src/VirtoCommerce.Platform.Data/GenericCrud/CrudService.cs [hunk 8]
```diff
@@ -0,0 +1,206 @@
...
+
+                }
+
+                return models;
+            });
+
+            return result.Select(x => (TModel)x.Clone());
+        }
+
+        protected virtual TModel AfterLoadEntities(string responseGroup, TEntity entity, TModel model)
```
Comment (Vectorfield4) [inline src/VirtoCommerce.Platform.Data/GenericCrud/CrudService.cs:75]:
```suggestion
        protected virtual TModel ProcessModel(string responseGroup, TEntity entity, TModel model)
```

File: src/VirtoCommerce.Platform.Data/GenericCrud/CrudService.cs [hunk 9]
```diff
@@ -0,0 +1,206 @@
...
+                    cacheEntry.AddExpirationToken(CreateCacheToken(ids));
+
+                    var entities = await LoadEntities(repository, ids, responseGroup);
+
+                    foreach (var entity in entities)
+                    {
+                        var model = entity.ToModel(AbstractTypeFactory<TModel>.TryCreateInstance());
+                        model = AfterLoadEntities(responseGroup, entity, model);
+                        if (model != null) models.Add(model);
+                    }
```
Comment (Vectorfield4) [inline src/VirtoCommerce.Platform.Data/GenericCrud/CrudService.cs:65]:
```suggestion
                    return entities.Select(entity => 
                    {
                        var model = entity.ToModel(AbstractTypeFactory<TModel>.TryCreateInstance());
                        return AfterLoadEntities(responseGroup, entity, model);
                    })
                    .Where(model => model != null);
```

Comment (Vectorfield4) [reply src/VirtoCommerce.Platform.Data/GenericCrud/CrudService.cs:65]: We don't need to create a list on L46.

File: src/VirtoCommerce.Platform.Data/GenericCrud/CrudService.cs [hunk 10]
```diff
@@ -0,0 +1,207 @@
...
+        }
+
+        protected abstract Task<IEnumerable<TEntity>> LoadEntities(IRepository repository, IEnumerable<string> ids, string responseGroup);
+
+        protected Task<IEnumerable<TEntity>> LoadEntities(IRepository repository, IEnumerable<string> ids)
+        {
+            return LoadEntities(repository, ids, "Full");
+        }
+
+        protected virtual Task BeforeSaveChanges(IEnumerable<TModel> models)
```
Comment (tatarincev) [inline src/VirtoCommerce.Platform.Data/GenericCrud/CrudService.cs:86]: why do not make these methods abstract?

File: src/VirtoCommerce.Platform.Data/GenericCrud/CrudService.cs [hunk 11]
```diff
@@ -0,0 +1,207 @@
...
+            // Basic implementation of soft delete intentionally left empty.
+            return Task.CompletedTask;
+        }
+
+        public virtual async Task SaveChangesAsync(IEnumerable<TModel> models)
+        {
+            var pkMap = new PrimaryKeyResolvingMap();
+            var changedEntries = new List<GenericChangedEntry<TModel>>();
+
+            BeforeSaveChanges(models);
```
Comment (tatarincev) [inline src/VirtoCommerce.Platform.Data/GenericCrud/CrudService.cs:109]: await is missed

File: src/VirtoCommerce.Platform.Data/GenericCrud/SearchService.cs [hunk 12]
```diff
@@ -0,0 +1,87 @@
...
+                    var needExecuteCount = criteria.Take == 0;
+
+                    if (criteria.Take > 0)
+                    {
+                        var ids = await query.OrderBySortInfos(sortInfos).ThenBy(x => x.Id)
+                                         .Select(x => x.Id)
+                                         .Skip(criteria.Skip).Take(criteria.Take)
+                                         .ToArrayAsync();
+
+                        result.TotalCount = ids.Count();
```
Comment (tatarincev) [inline src/VirtoCommerce.Platform.Data/GenericCrud/SearchService.cs:58]: use CountAsync() instead

**Review by tatarincev — CHANGES_REQUESTED** [review]

File: src/VirtoCommerce.Platform.Data/GenericCrud/CrudService.cs (hunk 10 above)
Comment (akak1977) [reply src/VirtoCommerce.Platform.Data/GenericCrud/CrudService.cs:86]: Think, there are services without such actions

**Review by Vectorfield4 — APPROVED** [review]
Is it possible to reduce the number of generic parameters?

**Comment by akak1977** [comment]
No clear way to reduce generic parameters

File: src/VirtoCommerce.Platform.Caching/GenericCrud/GenericCrudCachingRegion.cs [hunk 13]
```diff
@@ -0,0 +1,40 @@
...
+using VirtoCommerce.Platform.Core.Caching;
+using VirtoCommerce.Platform.Core.Common;
+
+namespace VirtoCommerce.Platform.Caching.GenericCrud
+{
+    /// <summary>
+    /// Generic CRUD caching region implementation for use with generic crud service.
+    /// </summary>
+    /// <typeparam name="T"></typeparam>
+    public class GenericCrudCachingRegion<T> : CancellableCacheRegion<GenericCrudCachingRegion<T>> where T : Entity
```
Comment (tatarincev) [inline src/VirtoCommerce.Platform.Caching/GenericCrud/GenericCrudCachingRegion.cs:14]: The current implementation is turned out  quite general, thereby I think we can rename `GenericCrudCachingRegion` -> `GenericCachingRegion` and move it to the https://github.com/VirtoCommerce/vc-platform/tree/master/src/VirtoCommerce.Platform.Caching

**Review by tatarincev — APPROVED** [review]

File: src/VirtoCommerce.Platform.Caching/GenericCrud/GenericCrudCachingRegion.cs (hunk 13 above)
Comment (akak1977) [reply src/VirtoCommerce.Platform.Caching/GenericCrud/GenericCrudCachingRegion.cs]: OK

**Pull request merged**